src/schedule/ucc_schedule.h

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+#ifndef UCC_SCHEDULE_H_
+#define UCC_SCHEDULE_H_
+
+#include "ucc/api/ucc.h"
+#include "utils/ucc_list.h"
+#define MAX_LISTENERS 8
+
+typedef enum {
+    UCC_EVENT_COMPLETED = 0,
+    UCC_EVENT_SCHEDULE_STARTED,
+    UCC_EVENT_LAST
+} ucc_event_t;
+
+typedef struct ucc_coll_task ucc_coll_task_t;
+
+typedef ucc_status_t (*ucc_task_event_handler_p)(ucc_coll_task_t *task);
+
+typedef struct ucc_event_manager {
+    ucc_coll_task_t *listeners[UCC_EVENT_LAST][MAX_LISTENERS];
+    int              listeners_size[UCC_EVENT_LAST];
+} ucc_event_manager_t;
+
+typedef struct ucc_coll_task {
+    ucc_coll_req_t             super;
+    ucc_event_manager_t        em;
```

Comment:
- Why have em for every task - may be have one event manager per schedule? Particularly, do you expect multiple handlers (in this case max 8) for each event per task? 

- Another approach would be to have a common event manager for all schedules and tasks. Is there a reason to not do this way?

Reply:
There is no such "other" approach actually if think about it and try writing the implementation on paper. This is well known "observer" pattern. In general each our task might have multiple dependencies AND it might depend on multiple other tasks as well. This is why each "task" has em - this way it tells other tasks (depending on it) that event happened.

Schedule is just a special kind of task - it subscribes to all other tasks in the schedule and basically accounting purposes (how many tasks from the schedule are completed).

Regarding first comment: yes currently MAX_TASK is constant for all the tasks even though most common case is just 1 or 2 listeners or even 0. See the comment above regarding making this "dynamic" and working for all possible number of listeners (imho does not worth bothering). currently task is 208 bytes. If i set MAX_LISTENERS to 4 it will be 144. Setting it less does not make any sense (in hierarchical bcast i will have task that 3 listeners). We can start with 4 and we always have the assertion in the code. So if someone will try to write an algorithm with more than 4 listeners (which is unlikely) we can consider increasing it. Or adding little bit "slower" path of dynamic listeners allocation for case like that. At any rate Cls and TLs will inherit from ucc_coll_task_t and i'm sure their additional request size will likely be larger.

Reply:
Yes, the task and dependencies/observer pattern is obvious for me. 

The question I was asking above was more of a leading question and also has two parts. I will ask differently: (1) How do you progress/process all tasks and schedules? If you have a centralized event manager (say per UCC context) you can look at all of the ready/completed tasks. With this approach, I'm not sure how you figure out the tasks that are ready during progress (Maybe I have to wait for the next set of PRs to answer this question. :)) (2) Having per task event manager might be expensive because either you overestimate or underestimate the listeners required. If you overestimate you waste memory, and if you underestimate (you might need dynamic allocation) as you suggest.

Reply:
As discussed in the WG, we will document the design.